src/katgpucbf/xbgpu/bsend.py

```diff
@@ -0,0 +1,401 @@
+    data
+        Storage for tied-array-channelised-voltage data, with shape (n_frames,
+        n_beams, n_channels_per_substream, n_spectra_per_heap, COMPLEX) and
+        dtype :const:`SEND_DTYPE`.
+    saturated
+        Storage for saturation counts, with shape (n_frames, n_beams) and dtype
+        uint32.
+    channel_offset
+        The first frequency channel processed.
+    timestamp_step
+        Timestamp step between successive :class:`Frame` in a chunk.
+    """
+
+    def __init__(
+        self,
+        data: np.ndarray,
+        saturated: np.ndarray,
+        *,
+        channel_offset: int,
+        timestamp_step: int,
+    ) -> None:
+        n_frames = data.shape[0]
+        self.data = data
+        self.saturated = saturated
+
+        self._timestamp = 0
+        self._timestamp_step = timestamp_step
+        self._timestamps = (np.arange(n_frames) * self._timestamp_step).astype(">u8")
+
+        # NOTE: The future indicates when it is safe to modify the chunk,
+        # i.e. it is not being transmitted. At construction there is nothing to
+        # wait for, so we mark it ready.
+        self.future = asyncio.get_running_loop().create_future()
+        self.future.set_result(None)
+
+        self._frames = [
+            Frame(
+                self._timestamps[i, ...],
+                data[i],
+                saturated[i],
+                channel_offset=channel_offset,
+            )
+            for i in range(n_frames)
+        ]
+
+    @property
+    def timestamp(self) -> int:
+        """
+        Timestamp of the first heap.
+
+        Setting this property updates the timestamps stored in all the heaps.
+        This should only be done when :attr:`future` is done.
+        """
+        return self._timestamp
+
+    @timestamp.setter
+    def timestamp(self, value: int) -> None:
+        delta = value - self.timestamp
+        self._timestamps += delta
+        self._timestamp = value
+
+    def send(
+        self,
+        send_stream: "BSend",
+        time_converter: TimeConverter,
+        sensors: SensorSet,
+    ) -> asyncio.Future:
+        """
+        Transmit a chunk's heaps over a SPEAD stream.
+
+        This method returns immediately and sends the data asynchronously. Before
+        modifying the chunk, first await :attr:`future`.
+
+        .. todo::
+
+            Also update its relevant counters and sensor values.
+        """
+        if any(send_stream.tx_enabled):
+            for frame in self._frames:
+                heaps_to_send = [heap for heap, enabled in zip(frame.heaps, send_stream.tx_enabled) if enabled]
+                send_future = send_stream.stream.async_send_heaps(heaps_to_send, mode=spead2.send.GroupMode.ROUND_ROBIN)
+            # TODO: Update counters and sensor with chunk.saturation
+            # NOTE: We use the last future returned above as the data is sent
+            # in-order. Once the last frame has been transmitted, the rest have
+            # too.
```

Comment:
I've just realised there is a flaw in this idea - sorry for leading you on a wild goose chase (again!) It's fine if there are no errors, but there are a couple of problems if any of the sends fail:
1. If the *last* send fails because there isn't enough space in the queue (which should be impossible, but as you know bugs can happen), it will fail immediately, even though previous heaps are still in flight. That could allow data to be modified before it's fully transmitted.
2. If an earlier send fails, nothing catastrophic happens but the error is silently lost.

So unfortunately I'm going to need to ask you change this back to using gather :-(.

Reply:
Ah, too right! Honestly, I was surprised by [your confidence in](https://github.com/ska-sa/katgpucbf/pull/650#discussion_r1372631225)
> [The data is sent in-order, so] once the last frame has been transmitted, the rest have too.

Genuinely, not a problem and appreciate revisiting this point.